src/Dataset.ts

```diff
@@ -0,0 +1,218 @@
+// A Dataset manages the production and manipulation of *tiles*.
+
+import { Tile, Rectangle, QuadTile } from './tile';
+import {
+  extent, range, min, max, bisectLeft,
+} from 'd3-array';
+import Zoom from './interaction';
+import * as Comlink from 'comlink';
+
+//@ts-ignore
+import TileWorker from './tileworker.worker.js?worker&inline';
+import { APICall } from './types';
+import Scatterplot from './deepscatter';
+type Key = string;
+
+export abstract class Dataset<T extends Tile> {
+  abstract root_tile : T;
+//  public mutations : 
+  public max_ix : number = -1;
+  protected plot : Scatterplot;
+  protected _tileworkers: TileWorker[] = [];
+  abstract ready : Promise<void>;
+  constructor(plot : Scatterplot) {
+    this.plot = plot;
+  }
+  static from_quadfeather(url : string, prefs: APICall, plot: Scatterplot) : QuadtileSet {
+    if (url.match(/(\/[0-9]+){3}/)) {
+      throw new Error('Quadfeathers must be loaded from a base URL.');
+    }
+    return new QuadtileSet(url, prefs, plot);
+  }
+
+  get extent() : Rectangle {
+    return {
+      x: [-1e16, 1e16],
+      y: [-1e16, 1e16],
+    }
+  }
+
+  download_most_needed_tiles(bbox : Rectangle, max_ix: number, queue_length : number) {
+    throw new Error('Tile download not implemented');
+  }
+
+  map(callback : (tile: T) => any, after = false) {
```

Comment:
I think generics could make this method strongly typed

Reply:
I think I understand what you're saying here--like this? 
```ts
  map<U extends any>(callback : (tile: T) => U, after = false) : U[] {
    // perform a function on each tile and return the values in order.
    // after: whether to perform the function in bottom-up order
    const results : U[] = [];
    this.visit((d : T) => { results.push(callback(d)); }, after = after);
    return results;
  }
```

Reply:
Thanks for these comment @DonIsaac , much appreciated.

Reply:
You're very welcome! This is correct, with one tweak: because `any` lacks _all_ type constraints, and the point of `T extends <blank>` is to constrain the possible values of `T`, using `T extends any` has no effect and is functionally equivalent to `T`.